src/crypto/aes.js

```diff
@@ -0,0 +1,89 @@
+import CID from '../cid.js'
+import random from 'js-crypto-random'
+import aes from 'js-crypto-aes'
+
+/**
+ * @param {number} value
+ */
+const enc32 = value => {
+  value = +value
+  const buff = new Uint8Array(4)
+  buff[3] = (value >>> 24)
+  buff[2] = (value >>> 16)
+  buff[1] = (value >>> 8)
+  buff[0] = (value & 0xff)
+  return buff
+}
+
+/**
+ * @param {Uint8Array} buffer
+ */
+const readUInt32LE = (buffer) => {
+  const offset = buffer.byteLength - 4
+  return ((buffer[offset]) |
+          (buffer[offset + 1] << 8) |
+          (buffer[offset + 2] << 16)) +
+          (buffer[offset + 3] * 0x1000000)
+}
+
+/**
+ * @param {Uint8Array[]} buffers
+ */
+const concat = buffers => Uint8Array.from(buffers.map(b => [...b]).flat())
+
+/**
+ * @template {'aes-gcm' | 'aes-cbc' | 'aes-ctr'} Name
+ * @template {number} Code
+ * @param {Object} options
+ * @param {Name} options.name
+ * @param {Code} options.code
+ * @param {number} options.ivsize
+ */
+const mkcrypto = ({ name, code, ivsize }) => {
+  // Line below does a type cast, because type checker can't infer that
+  // `toUpperCase` will result in desired string literal.
+  const cyperType = /** @type {import('js-crypto-aes/dist/params').cipherTypes} */(name.toUpperCase())
+  /**
+   * @param {Object} options
+   * @param {Uint8Array} options.key
+   * @param {Object} options.value
+   * @param {Uint8Array} options.value.bytes
+   * @param {Uint8Array} options.value.iv
+   */
+  const decrypt = async ({ key, value: { iv, bytes } }) => {
```

Comment:
It appears to me that it would make a lot of sense to give `{iv, bytes, code}` struct a name (and a type) so it's easier to conceptualize and talk about it.

I would also expect `decrypt` to expect `value.code` and assert it against the `code` instead of just ignoring it all together.

That way decrypt turns `key+namedStruct` into `cid+bytes` and encrypt turns `key+cid-bytes` into `namedStruct`.

Odd obsession with encrypt/decrypt symmetry also makes me wonder keys should be included in return types as well which would make output of encrypt input of decrypt and vice versa.

Reply:
I think there’s more utility in having the implementation *just* use this destructing because then you don’t need to load the type in order to call the method.

If i know that I’m going to call the “decrypt AES function” I can call this with:

```js
const iv = Buffer.from(something1)
const bytes = Buffer.from(something2)
const key = Buffer.from(something3)
decrypt({ key, value: { iv, bytes }})
```

But also, if I just have a block and I know that it’s encrypted I can destructure it into the call.

```js
const key = Buffer.from(something)
descrypt({ key, ...block })
```

This makes the decrypt function’s interface work with **any** block based decryption function. It may or may not have an `iv`, it may have other information in the value, cid, or bytes of the block it wants, and all of that will “just work” if I’m
using block destructuring.